### PR TITLE
[SPARK-35100][ML] Refactor AFT - support virtual centering

### PR DESCRIPTION
**Pull request by zhengruifeng**
### What changes were proposed in this pull request?
1, remove old agg;
2, apply new agg supporting virtual centering

### Why are the changes needed?
for better convergence

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
existing testsuites

### CODE REVIEW

**Comment by Yikun** [comment]
Friendly note: we process the fast forward error on https://github.com/apache/spark/pull/32182 , and you could just sync your  `zhengruifeng:refactor_aft` with upstream to make the build and test work.

**Review by srowen — COMMENTED** [review]
Looks generally fine - any opportunities to refactor some common block aggregator code now that there are many of them?

File: mllib/src/main/scala/org/apache/spark/ml/optim/aggregator/AFTBlockAggregator.scala [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+      case dm: DenseMatrix =>
+        BLAS.nativeBLAS.dgemv("N", dm.numCols, dm.numRows, 1.0, dm.values, dm.numCols,
+          arr, 1, 1.0, gradientSumArray, 1)
+
+      case sm: SparseMatrix =>
+        val linearGradSumVec = new DenseVector(Array.ofDim[Double](numFeatures))
+        BLAS.gemv(1.0, sm.transpose, vec, 0.0, linearGradSumVec)
+        BLAS.javaBLAS.daxpy(numFeatures, 1.0, linearGradSumVec.values, 1,
+          gradientSumArray, 1)
+    }
```
Comment (zhengruifeng) [inline mllib/src/main/scala/org/apache/spark/ml/optim/aggregator/AFTBlockAggregator.scala:126]: > any opportunities to refactor some common block aggregator code now that there are many of them?

I think it is possible to simplify here by adding some BLAS util fuctions in the future. Then we do not need to create some temporary vector/matrix/array for just matching the shape. I just have a idea now, and will check it in the future. If possible, there will be a seperate PR.

**Comment by zhengruifeng** [comment]
retest this please

**Comment by zhengruifeng** [comment]
@srowen I think this PR is ready. Would you mind making a final review? As to refactoring some common block aggregator code, I will try to do this after LiR get changed in this way.

**Review by srowen — COMMENTED** [review]
Should this wait until other BLAS related changes to make sure it's consistent? not sure if it is a problem. The approach looks fine, if it's consistent with how you changed other aggregators.

**Comment by zhengruifeng** [comment]
@srowen It should be safe to wait for BLAS related changes. I will keep this PR open.

**Comment by zhengruifeng** [comment]
retest this please

**Comment by zhengruifeng** [comment]
friendly ping @srowen , I think it is ready to merge

**Review by srowen — APPROVED** [review]

**Comment by zhengruifeng** [comment]
Thanks @srowen for reviewing! I will make LinearRegression support virtual centering and then refactor some common blocks related to BLAS to simplify the impls.

**Comment by zhengruifeng** [comment]
Merged to master

**Pull request closed without merging**